hail/src/main/scala/is/hail/asm4s/Code.scala

```diff
@@ -158,15 +158,15 @@ object Code {
     }
   }
 
-  def whileLoop(cond: Code[Boolean], body: Code[_]*): Code[Unit] = {
-    import is.hail.asm4s.joinpoint._
-    JoinPoint.CallCC[Unit] { (jb, break) =>
-      val continue = jb.joinPoint()
-      val loopBody = jb.joinPoint()
-      continue.define { _ => JoinPoint.mux(cond, loopBody, break) }
-      loopBody.define { _ => Code(Code(body: _*), continue(())) }
-      continue(())
-    }
+  def whileLoop(cond: Code[Boolean], body: Code[Unit]*): Code[Unit] = {
+    val L = new CodeLabel()
+    Code(
+      L,
+      cond.mux(
+        Code(
+          Code(body: _*),
+          L.goto),
```

Comment:
I'd like to understand how you think about this in relation to the semantics "a `Code[T]` is something that leaves a `T` on the top of the stack when it finishes". The `Code(..., L.goto)` doesn't seem to fit that model. In particular, `L.goto` is different from other `Code[Unit]` because `Code(L.goto, ...)` isn't well formed.

Reply:
> because Code(L.goto, ...) isn't well formed

Sure it is:

```
Code(L.goto, L, more code)
```

Reply:
Okay, sure. But it is true that `L.goto` can't be used everywhere a `Code[Unit]` can be. `Code(a := 1, b := 2)` is fine, but `Code(L.goto, b := 2)` is not.

Reply:
I will flesh out a response to this.

Reply:
> I will flesh out a response to this.

Please see: https://dev.hail.is/t/proposal-low-level-ir/182

 > but Code(L.goto, b := 2) is not

I'd say this is fine, it is just dead code.

What about `Code(L.goto, 5)`?  I'm not 100% sure how the JVM would deal with it, but I bet it is also fine if the subsequent emitted code consumes b as intended.  Just more dead code.  So I think from this perspective `L.goto: Code[Unit]` is totally consistent.  However, I know you want more static checking.  That's included in my proposal above.